tests/bsim/bluetooth/host/settings/cleanup/src/test.c

```diff
@@ -0,0 +1,215 @@
+
+const char *expected_settings_key[] = {
+	"hash",
+	"id",
+	"sc/0000000000000",
+	"cf/0000000000000",
+	"ccc/0000000000000",
+	"sc/0000000000000/2",
+	"cf/0000000000000/2",
+	"ccc/0000000000000/2",
+	"mesh/dummy/key",
+};
+
+static int print_all_settings_cb(const char *key, size_t len, settings_read_cb read_cb,
+				 void *cb_arg, void *param)
+{
+	ssize_t err;
+	uint8_t *data = k_malloc(len);
+
+	err = read_cb(cb_arg, data, len);
+	if (err != len) {
+		FAIL("Failed to read data (err %d)\n", err);
+	}
+
+	LOG_INF("key: 'bt/%s'", key);
+	LOG_HEXDUMP_INF(data, len, "value:");
+
+	k_free(data);
+
+	return 0;
+}
+
+static void print_all_settings(void)
+{
+	int err;
+
+	err = settings_load_subtree_direct("bt", print_all_settings_cb, NULL);
+	if (err) {
+		FAIL("Failed to load 'bt' subtree (err %s)\n", err);
+	}
+}
+
+static int check_settings_cb(const char *key, size_t len, settings_read_cb read_cb, void *cb_arg,
+			     void *param)
+{
+	bool key_is_expected = false;
+	uint8_t *key_not_deleted = (uint8_t *)param;
+
+	for (size_t i = 0; i < ARRAY_SIZE(expected_settings_key); i++) {
+		if (strcmp(expected_settings_key[i], key) == 0 && key_not_deleted[i] == 0) {
+			key_not_deleted[i] = 1;
+			key_is_expected = true;
+		}
+	}
+
+	if (!key_is_expected) {
+		FAIL("Key 'bt/%s' should have been deleted\n", key);
+	}
+
+	return 0;
+}
+
+static void check_settings(void)
+{
+	int err;
+	uint8_t key_not_deleted[ARRAY_SIZE(expected_settings_key)] = {0};
+
+	err = settings_load_subtree_direct("bt", check_settings_cb, key_not_deleted);
+	if (err) {
+		FAIL("Failed to load 'bt' subtree (err %d)\n", err);
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(expected_settings_key); i++) {
+		if (key_not_deleted[i] == 0) {
+			FAIL("Key '%s' should not have been deleted\n", expected_settings_key[i]);
+		}
+	}
+}
+
+static void populate_settings(void)
+{
+	int err;
+	uint8_t dummy_value = 0;
+	bt_addr_le_t dummy_addr;
+	bt_addr_le_t ids_addr[CONFIG_BT_ID_MAX] = {0};
+
+	__ASSERT_NO_MSG(CONFIG_BT_ID_MAX >= 3);
+
+	bt_addr_le_copy(&dummy_addr, BT_ADDR_LE_ANY);
+
+	err = bt_addr_le_create_static(&ids_addr[0]);
+	if (err) {
+		FAIL("Failed to create static address (err %d)\n", err);
+	}
+
+	err = bt_addr_le_create_static(&ids_addr[2]);
+	if (err) {
+		FAIL("Failed to create static address (err %d)\n", err);
+	}
+
+	err = settings_save_one("bt/id", ids_addr, sizeof(ids_addr));
+	if (err) {
+		FAIL("Failed to save ID (err %d)\n", err);
+	}
+
+	bt_settings_store_sc(0, &dummy_addr, &dummy_value, sizeof(dummy_value));
+	bt_settings_store_cf(0, &dummy_addr, &dummy_value, sizeof(dummy_value));
+	bt_settings_store_ccc(0, &dummy_addr, &dummy_value, sizeof(dummy_value));
+
+	bt_settings_store_sc(1, &dummy_addr, &dummy_value, sizeof(dummy_value));
+	bt_settings_store_cf(1, &dummy_addr, &dummy_value, sizeof(dummy_value));
+	bt_settings_store_ccc(1, &dummy_addr, &dummy_value, sizeof(dummy_value));
+
+	settings_save_one("bt/dummy/key", &dummy_value, sizeof(dummy_value));
+
```

Comment:
There is a bug in the DUT due to how `strncmp` is used. 

```suggestion
	settings_save_one("bt/i", &dummy_value, sizeof(dummy_value));
```

Reply:
Actually that's a bug in `bt_settings_load_direct_cleanup`, thanks for catching it!

When testing if a key is known or not, here:
https://github.com/zephyrproject-rtos/zephyr/blob/03e3c73ee32c0a3c536a4657dfe171163f05128e/subsys/bluetooth/host/settings.c#L337-L341

I compare the two strings with the size of the one we want to test. That's wrong, I should compare with the size of the string in `bt_settings_keys`. Because here it compare with `bt/id` but with the size of `bt/i`, so it say that the key is known.

I'll fix that and add the test.

Reply:
Now it's wrong the other way. Comparing strings in C is hard :P

```suggestion
settings_save_one("bt/idd", &dummy_value, sizeof(dummy_value));
```

Reply:
Indeed, it should be better now. If `strncmp` say that the string are equal I then check the expected last character of the tested key to see if it's `\0` or `/`.